README
NAME
    PPIx::IndexLines - Given a line number, returns some basic information
    about where in the perl document you are

VERSION
    version 0.03

SYNOPSIS
      #!/usr/bin/perl
      use strict;
      use warnings;
      use PPIx::IndexLines;
      my $document = PPIx::IndexLines->new( +shift );
      $document->index_lines;
      print $document->line_type( +shift );

METHODS
  new
    Accepts either a scalar or a scalar reference. This is passed directly
    to PPI::Document::new, so the same rules as in that method apply.

    Basically, if you pass in a scalar, it will be assumed to be a filename
    and said file will be loaded.

    Otherwise, a scalar reference is assumed to be a PPI document and will
    be parsed directly.

  index_lines
    Call this method to index the lines for the new PPI::Document.

  line_type
    Call this method with a line number, it will return one of

      BEGIN CHECK UNITCHECK INIT END POD __DATA__ __END__

    or

      <PackageName>::<SubName>

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

SEE ALSO
    *   PPIx::IndexOffsets

    *   PPIx::LineToSub

AUTHOR
    Alan Young <harleypig@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2010 by Alan Young.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Revision History for PPIx-IndexLines
====================================

v0.03 2010-12-09
----------------

commit d913ea1a3c5d5d51e16040e22460c5e9d459bb6f
Author: Harley Pig <harleypig@gmail.com>

    final commit before release

v0.02 2010-12-07
----------------

commit d9c0aa605c6c48b3874b89254eacd94a1e2463b5
Author: Harley Pig <harleypig@gmail.com>

    making sure changes are committed

v0.01 2010-11-29
----------------

commit e31a7d1dee59c004728bfcac128912bb8bf55d02
Author: Harley Pig <harleypig@gmail.com>

    nearly done

commit e8c96bee1a247c7967d4cef64f0a8bff611043c9
Author: Harley Pig <harleypig@gmail.com>

    making things work

===================================
End of changes in the last 365 days
===================================


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  'build_requires' => {
    'Carp' => '0',
    'English' => '0',
    'File::Find' => '0',
    'File::Temp' => '0',
    'Module::Build' => '0.3601',
    'Path::Class' => '0',
    'Scalar::Util' => '0',
    'Test::More' => '0.94',
    'Test::Perl::Critic::Progressive' => '0',
    'Try::Tiny' => '0',
    'perl' => '5.004'
  },
  'configure_requires' => {
    'ExtUtils::MakeMaker' => '6.31',
    'Module::Build' => '0.3601'
  },
  'dist_abstract' => 'Given a line number, returns some basic information about where in the perl document you are',
  'dist_author' => [
    'Alan Young <harleypig@gmail.com>'
  ],
  'dist_name' => 'PPIx-IndexLines',
  'dist_version' => '0.03',
  'license' => 'perl',
  'module_name' => 'PPIx::IndexLines',
  'recommends' => {},
  'recursive_test_files' => 1,
  'requires' => {
    'PPI' => '0'
  },
  'script_files' => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/PPIx/IndexLines.pm
package PPIx::IndexLines;

# ABSTRACT: Given a line number, returns some basic information about where in the perl document you are

use strict;

use PPI;

# VERSION

=method new

Accepts either a scalar or a scalar reference.  This is passed directly to
PPI::Document::new, so the same rules as in that method apply.

Basically, if you pass in a scalar, it will be assumed to be a filename and
said file will be loaded.

Otherwise, a scalar reference is assumed to be a PPI document and will be
parsed directly.

=cut

sub new { PPI::Document->new( $_[ 1 ] ) }

=method index_lines

Call this method to index the lines for the new PPI::Document.

=cut

sub index_lines {

  my $self = shift;

  my @el = $self->elements;

  my $package = 'main';
  my $info;

  my %lookup = (

    'PPI::Statement::Data'      => sub { '__DATA__' },
    'PPI::Statement::End'       => sub { '__END__' },
    'PPI::Statement::Package'   => sub { $package = $_[ 0 ]->namespace },
    'PPI::Statement::Scheduled' => sub { $_[ 0 ]->type },
    'PPI::Statement::Sub'       => sub { sprintf '%s::%s', $package, $_[ 0 ]->name },

    # Is this interesting enough to include?
    #'PPI::Token::Comment'       => sub { sprintf '%s (%s)', $package, 'comment' },

    'PPI::Token::Pod' => sub { 'POD' },

  );

  my $last_line_number = 0;

  for my $e ( @el ) {

    my $line_number = $e->line_number;

    next if exists $self->{ 'line' }{ $line_number };

    $self->{ 'line' }{ $_ } = $info for $last_line_number + 1 .. $line_number - 1;

    $last_line_number = $line_number;

    my $node_type = ref $e;

    $self->{ 'line' }{ $line_number } = $info = exists $lookup{ $node_type } ? $lookup{ $node_type }->( $e ) : $package;

  }
} ## end sub index_lines

=method line_type

Call this method with a line number, it will return one of

  BEGIN CHECK UNITCHECK INIT END POD __DATA__ __END__

or

  <PackageName>::<SubName>

=cut

sub line_type { $_[ 0 ]->{ 'line' } || '' }

1;

=head1 SYNOPSIS

  #!/usr/bin/perl
  use strict;
  use warnings;
  use PPIx::IndexLines;
  my $document = PPIx::IndexLines->new( +shift );
  $document->index_lines;
  print $document->line_type( +shift );

=head1 SEE ALSO

L<PPIx::IndexOffsets>
L<PPIx::LineToSub>

=cut


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Given a line number, returns some basic information about where in the perl document you are',
  'AUTHOR' => 'Alan Young <harleypig@gmail.com>',
  'BUILD_REQUIRES' => {
    'Carp' => '0',
    'English' => '0',
    'File::Find' => '0',
    'File::Temp' => '0',
    'Module::Build' => '0.3601',
    'Path::Class' => '0',
    'Scalar::Util' => '0',
    'Test::More' => '0.94',
    'Test::Perl::Critic::Progressive' => '0',
    'Try::Tiny' => '0',
    'perl' => '5.004'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31',
    'Module::Build' => '0.3601'
  },
  'DISTNAME' => 'PPIx-IndexLines',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'PPIx::IndexLines',
  'PREREQ_PM' => {
    'PPI' => '0'
  },
  'VERSION' => '0.03',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



